Adding non-standard ways to use EESSI (#387)
EESSI/docs: docs/getting_access/non_standard.md

    # cleanup
    rm -rf $tmpdir
fi

```

Do make sure that this script is executable:

```bash
    chmod u+x ~/bin/cvmfsexec_eessi.sh
```

A simple way to test this script is to use it to inspect the contents of the EESSI repository:

```bash
    ~/bin/cvmfsexec_eessi.sh ls /cvmfs/software.eessi.io
```

or to start an interactive shell in which the EESSI repository is mounted:

```bash
    # Expected output from a machine without a cvmfs installation
    $ ls /cvmfs
    ls: cannot access '/cvmfs': No such file or directory

    # Starting interactive shell
    $ ~/bin/cvmfsexec_eessi.sh /bin/bash -l
    CernVM-FS: loading Fuse module... done
    CernVM-FS: mounted cvmfs on /tmp/hvela/tmp.iH6e993Adw/cvmfsexec/dist/cvmfs/cvmfs-config.cern.ch
    CernVM-FS: loading Fuse module... done
    CernVM-FS: mounted cvmfs on /tmp/hvela/tmp.iH6e993Adw/cvmfsexec/dist/cvmfs/software.eessi.io

    # Now it's mounted!
    $ ls /cvmfs/
    cvmfs-config.cern.ch  software.eessi.io

```

Notice how now that EESSI is mounted, you will need to [set up the environment](https://www.eessi.io/docs/using_eessi/setting_up_environment/) in order to access the software itself.

### `orted`  wrapper script

> In order to get multi-node runs of software working without having EESSI available system-wide, we also had to create a small wrapper script for the `orted` command that is used by Open MPI to start processes on remote nodes. This is necessary because `mpirun` launches `orted`, which must be run in an environment in which the EESSI repository is mounted. If not, MPI startup will fail with an error like

Reviewer: ```suggestion
In order to get multi-node runs of software working without having EESSI available system-wide, you also need to create a small wrapper script for the `orted` command that is used by Open MPI to start processes on remote nodes. This is necessary because `mpirun` launches `orted`, which must be run in an environment in which the EESSI repository is mounted. If not, MPI startup may fail with an error like
```

```bash
"error: execve(): orted: No such file or directory".
```

This wrapper script must be named `orted`, and must be located in a path that is listed in `$PATH`.

> We placed it in `~/bin/orted`, and you can add `export PATH=$HOME/bin:$PATH` to your `~/.bashrc` login script.

Reviewer: ```suggestion
For example, it can be placed in `~/bin/orted`, and you can add `export PATH=$HOME/bin:$PATH` to your `~/.bashrc` login script.
```

Contents of ~/bin/orted:

```bash
#!/bin/bash

# first remove path to this orted wrapper from $PATH, to avoid infinite loop
orted_wrapper_dir=$(dirname $0)
export PATH=$(echo $PATH | tr ':' '\n' | grep -v $orted_wrapper_dir | tr '\n' ':')

> ~/bin/cvmfsexec_eessi.sh orted "$@"

Reviewer: Does this work? I imagine that you need to load the modules you need in your `cvmfsexec_eessi.sh` script or it will not find `orted` inside the container

Reviewer 2: I think the reason this works is that the context _calling_ the `orted` command already has the module loaded, isn't it? I.e. the typical usage is:

```
module load EESSI/2023.06
module load GROMACS/whatever
# This mpirun calls `orted` on the remote node, probably exporting the _current_ `$PATH`, on which `orted` is available because the mpi module is loaded through GROMACS. It would fail if we _don't_ wrap that `orted` command in a `~/bin/cvmfsexec_eessi.sh`, because on the remote node the CVMFS repo is _not_ yet mounted. However, mounting the CVMFS repo is _sufficient_, because the `$PATH` is correct - it just points to the CVMFS repo...
mpirun <something>
```

```

Again, do make sure that also this `orted` wrapper script is executable:

```bash
chmod u+x ~/bin/orted
```

If not, you will likely run into an error that starts with:

```bash
An ORTE daemon has unexpectedly failed after launch ...
```

### Slurm job script

The `cvmfsexec_eessi.sh` can be used inside a Slurm job script inside your HPC system to initialize the EESSI environment in a subshell which the EESSI CernVM-FS repository is mounted. Thus, you will be able to load any module from EESSI you might need.

Example job script:

```bash
#!/bin/bash
#SBATCH --ntasks=4
#SBATCH --ntasks-per-node=2
#SBATCH --cpus-per-task=1
#SBATCH --time=5:0:0
#SBATCH --export=None
#SBATCH --mem=30000M
~/bin/cvmfsexec_eessi.sh << EOF
source /cvmfs/software.eessi.io/versions/2023.06/init/bash
module load TensorFlow/2.13.0-foss-2023a
export SLURM_EXPORT_ENV=HOME,PATH,LD_LIBRARY_PATH,PYTHONPATH
mpirun -np 4 python test.py
EOF

```
You can see the original blog post on how they used this solution in Deucalion [here](https://www.eessi.io/docs/blog/2024/06/28/espresso-portable-test-run-eurohpc/#running-espresso-on-deucalion-via-eessi-cvmfsexec). 

## Via `squashfs` +  cvmfs's `shrinkwrap` utility

> CernVM-FS provides the Shrinkwrap utility, allowing users to create a portable snapshot of a CVMFS repository. This can be exported and distributed without the need of a CVMFS client or network access.

Reviewer: Should link to the CVMFS docs on this early on.

To create an export of EESSI in user space, you will first need to create the config file `software.eessi.io.config`:

```bash
CVMFS_REPOSITORIES=software.eessi.io
CVMFS_REPOSITORY_NAME=software.eessi.io
CVMFS_CONFIG_REPOSITORY=cvmfs-config.cern.ch
CVMFS_SERVER_URL='http://aws-eu-west-s1-sync.eessi.science/cvmfs/software.eessi.io'
CVMFS_HTTP_PROXY=DIRECT # Avoid filling up any local squid's cache
CVMFS_CACHE_BASE=/tmp/shrinkwrap
CVMFS_KEYS_DIR=/etc/cvmfs/keys/eessi.io # Need to be provided for shrinkwrap
CVMFS_SHARED_CACHE=no # Important as libcvmfs does not support shared caches
CVMFS_USER=cvmfs
CVMFS_UID_MAP=uid.map
CVMFS_GID_MAP=gid.map

```

> You will need to create the files `uid.map` and `gid.map` with the respective value you will use preceded by a `*`. P.e., assuming UID 1000, set the two following files:

Reviewer: ```suggestion
You will need to create the files `uid.map` and `gid.map` with the respective value you will use preceded by a `*`. For example, assuming UID 1000, set the two following files:
```

> You will need to create the files `uid.map` and `gid.map` with the respective value you will use preceded by a `*`. For example, assuming UID 1000, set the two following files:

Reviewer: I would explain to people that they _probably_ want to use the current user id and primary group, so
```bash
$ echo '*' $(id -u $USER) > uid.map
```
and
```bash
$ echo '*' $(id -g $USER) > uid.map
```

```bash
$ cvmfs_shrinkwrap -r software.eessi.io -f software.eessi.io.config -t software.eessi.io.spec --dest-base /tmp/cvmfs -j 4

LibCvmfs version 2.12, revision 31
(libcvmfs) (manager 'standard') switching proxy from (none) to DIRECT. Reason: set random start proxy from the first proxy group [Current host: http://aws-eu-west-s1-sync.eessi.science/cvmfs/software.eessi.io]
(libcvmfs) (manager 'external') switching proxy from (none) to DIRECT. Reason: cloned [Current host: http://aws-eu-west-s1-sync.eessi.science/cvmfs/software.eessi.io]
(libcvmfs) Starting 4 workers
(libcvmfs) cntByte|0|Byte count of projected repository
cntDir|1|Number of directories from source repository
cntFile|0|Number of files from source repository
cntSymlink|0|Number of symlinks from source repository
dataBytes|0|Bytes transferred from source to destination
dataBytesDeduped|0|Number of bytes not copied due to deduplication
dataFiles|0|Number of data files transferred from source to destination
dataFilesDeduped|0|Number of files not copied due to deduplication
entriesDest|1|Number of file system entries processed in the destination
entriesSrc|1|Number of file system entries processed in the source

....

> # This takes a long time

Reviewer: Should also mention that it takes a lot of space as well depending on how much you export.

```

Once completed, the contents will be available in /tmp/cvmfs. You can create an squashfs image from it:

```bash

>     sudo mksquashfs /tmp/cvmfs software.eessi.io.sqsh

Reviewer: sudo should not be necessary

Author: It shouldn't, but I got this same error in two separate systems, one with cvmfs already installed and the other one with `cvmfs_shrinkwrap `only, is anyone else experiencing it? 

```
(libcvmfs) Failed to create link : /tmp/cvmfs/.data/e8/ec/3d88b62ebf526e4e5a4ff6162a3aa48a6b78.713e9bb163f52062c01fa6a5d36c8e06->/tmp/cvmfs/software.eessi.io/versions/2023.06/software/linux/x86_64/intel/skylake_avx512/software/zstd/1.5.5-GCCcore-13.2.0/.cvmfscatalog : 13 : Permission denied

(libcvmfs) Failed to link /versions/2023.06/software/linux/x86_64/intel/skylake_avx512/software/zstd/1.5.5-GCCcore-13.2.0/.cvmfscatalog->/e8/ec/3d88b62ebf526e4e5a4ff6162a3aa48a6b78.713e9bb163f52062c01fa6a5d36c8e06 : 13 : Permission denied
(libcvmfs) File /versions/2023.06/software/linux/x86_64/intel/skylake_avx512/software/zstd/1.5.5-GCCcore-13.2.0 failed to copy
```

Reviewer: Did you use your uid/gid when creating the CVMFS export?

Reviewer: (you may need to completely remove `/tmp/cvmfs` to change uid/gid)

Author: It was a bug and very kindly the CVMFS fixed in the latest nightly release, working fine now!

```

This squashfs image can be mounted in any container:

```bash
    apptainer shell -B software.eessi.io.sqsh:/cvmfs:image-src=/ docker://ubuntu

```

> Right now its a manual process but there is work in progress towards an automated solution.

Reviewer: ```suggestion
Right now, this may seem like quite a manual process, but there is work in progress towards creating a script to automate this process.
```

To create an export of EESSI in user space, you will first need to create the config file `software.eessi.io.config`:

```bash
CVMFS_REPOSITORIES=software.eessi.io
CVMFS_REPOSITORY_NAME=software.eessi.io
CVMFS_CONFIG_REPOSITORY=cvmfs-config.cern.ch
CVMFS_SERVER_URL='http://aws-eu-west-s1-sync.eessi.science/cvmfs/software.eessi.io'
CVMFS_HTTP_PROXY=DIRECT # Avoid filling up any local squid's cache
CVMFS_CACHE_BASE=/tmp/shrinkwrap
CVMFS_KEYS_DIR=/etc/cvmfs/keys/eessi.io # Need to be provided for shrinkwrap
CVMFS_SHARED_CACHE=no # Important as libcvmfs does not support shared caches
CVMFS_USER=cvmfs
CVMFS_UID_MAP=uid.map
CVMFS_GID_MAP=gid.map

```
You will need to create the files `uid.map` and `gid.map` with the respective value you will use preceded by a `*`. You will probably want to use the UID and GID of the current user, so, to set the two files:

```bash
    $ echo '*' $(id -u $USER) > uid.map
    $ cat uid.map
    * 1001

    $ echo '*' $(id -g $USER) > gid.map
    $ cat gid.map
    * 1001
```

> In addition, you need to create a spec file `software.eessi.io.spec` with the files you want to include and/or exclude in the shrinkwrap. Contents are:

Reviewer 2: ```suggestion
In addition, you need to create a spec file `software.eessi.io.spec` with the files you want to include and/or exclude in the shrinkwrap. For example:
```
